design/20190708.certificate-request-crd.md

```diff
@@ -174,6 +188,190 @@ type ObjectReference struct {
+are mutually exclusive (i.e. a CertificateRequest cannot have an Approved and
+Denied condition simultaneously). This behaviour is enforced in the cert-manager
+validating admission webhook.
+
+An "approver" is an entity that is responsible for setting the Approved/Denied
+conditions. It is up to the approver's implementation as to what
+CertificateRequests are managed by that approver.
+
+The Reason field of the Approved/Denied condition should be set to *who* set the
+condition. Who can be interpreted however makes sense to the approver
+implementation. For example, it may include the API group of an approving policy
+controller, or the client agent of a manual request.
+
+The Message field of the Approved/Denied condition should be set to *why* the
+condition is set. Again, why can be interpreted however makes sense to the
+implementation of the approver. For example, the name of the resource that
+approves this request, the violations which caused the request to be denied, or
+the team to who manually approved the request.
+
+When a CertificateRequest has been Denied, it is the responsibility of the
+referenced issuer to then add a Ready condition with the status of "False",
+along with a relevant Reason and Message.
+
+
+##### RBAC
+
+Approved and Denied conditions are set by requesting against the `/status`
+endpoint of the CertificateRequest resource. This is a divergence of the
+certificates.k8s.io CertificateSigningRequest resource that utilises `approval`,
+as CustomResourceDefinitions may [only define `/status` or `/scale`
+sub-resources](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#subresources).
+Approvers must therefore have permissions to update the status sub-resource of
+CertificateRequests. Approvers missing this permissions will have their request
+rejected by the API server.
+
+Setting the Approved or Denied conditions are restricted by the approver having
+sufficient RBAC permissions. These permissions are based upon the request
+itself- specifically the request's IssuerRef:
+
+```yaml
+apiGroups: ["cert-manager.io"]
+resources: ["signers"]
+verbs: ["approve"]
+resourceNames:
+ # namesapced signers
+ - "<signer-resource-name>.<signer-group>/<signer-namespace>.<signer-name>"
+ # cluster scoped signers
+ - "<signer-resource-name>.<signer-group>/<signer-name>"
+ # all signers of this resource name
+ - "<signer-resource-name>.<signer-group>/*"
+```
+
+An example ClusterRole that would grant the permissions to set the Approve and
+Denied conditions of CertificateRequests that reference the cluster scoped
+`myissuers` external issuer, in the group `my-example.io`, with the name `myapp`:
+
+```
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: my-example-io-my-issuer-myapp-approver
+rules:
+  - apiGroups: ["cert-manager.io"]
+    resources: ["signers"]
+    verbs: ["approve"]
+    resourceNames: ["myissuers.my-example.io/myapp"]
+```
```

Comment:
Would be nice to use concrete names in the example to help the reader understand what this new RBAC syntax entails. I also propose to add a concrete example:

~~~suggestion

Let us imagine that we want to build an approver that uses OPA (Open Policy Agent). Our approver
takes the form of a Kubernetes controller that watches the CertificateRequest objects. For every
CertificateRequest, our approver asks the OPA engine whether the CertificateRequest should
be accepted or not:

- If OPA returns true, the approver adds the `Approved=True` condition to the CertificateRequest;
- If OPA returns false, the approver adds the `Denied=True` condition to the CertificateRequest.

By default, our approver does not have the permission to add the `Approved=True`
and `Denied=True` conditions. In order to allow it to add these conditions, we use a custom
RBAC syntax that cert-manager knows about. Let us imagine that our approver needs to approve
CertificateRequests that are targeting the internal ACME issuer as well as the external issuer for
[AWS PrivateCA](https://github.com/jniebuhr/aws-pca-issuer):

```
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: letsencrypt-issuer
  namespace: some-namespace
spec:
  acme:
    email: jane.doe@gmail.com
    server: https://acme-staging-v02.api.letsencrypt.org/directory
    privateKeySecretRef:
      name: letsencrypt-issuer-account
    solvers:
      - http01:
          ingress:
            class: traefik
---
apiVersion: awspca.cert-manager.io
kind: AWSPCAClusterIssuer
metadata:
  name: awspca-issuer
  namespace: some-namespace
spec:
  arn: "pca-807911e9"
```

To make our approver work for all CertificateRequests across all namespaces,
we apply the following role to our cluster:

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: opa-approver-for-acme-and-aws-pca
rules:
  - apiGroups: ["cert-manager.io"]
    resources: ["signers"]
    verbs: ["approve"]
    resourceNames:
      - "clusterissuers.acme.cert-manager.io/letsencrypt-issuer"
      #  <-------------------------------->  <---------------->
      #    resource type + resource group        issuer name
      - "awspcaclusterissuers.awspca.cert-manager.io/awspca-issuer"
```

Note: the "resource type" [is](https://kubernetes.io/docs/reference/using-api/api-concepts/#standard-api-terminology)
the lower-case and plural version of the "kind" field.
~~~

Reply:
I think this would be a good addition to the website, where if someone wanted to write an approver.